river/tree/splitter/nominal_splitter_classif.py

```diff
@@ -23,72 +29,61 @@ def update(self, att_val, target_val, sample_weight):
         if att_val is None:
             self._missing_weight_observed += sample_weight
         else:
-            try:
-                self._att_val_dist_per_class[target_val]
-            except KeyError:
-                self._att_val_dist_per_class[target_val] = {att_val: 0.0}
-                self._att_val_dist_per_class = dict(
-                    sorted(self._att_val_dist_per_class.items())
-                )
-            try:
-                self._att_val_dist_per_class[target_val][att_val] += sample_weight
-            except KeyError:
-                self._att_val_dist_per_class[target_val][att_val] = sample_weight
-                self._att_val_dist_per_class[target_val] = dict(
-                    sorted(self._att_val_dist_per_class[target_val].items())
-                )
+            self._att_values.add(att_val)
+            self._att_dist_per_class[target_val][att_val] += sample_weight
 
         self._total_weight_observed += sample_weight
 
-        return self
-
     def cond_proba(self, att_val, target_val):
-        obs = self._att_val_dist_per_class.get(target_val, None)
-        if obs is not None:
-            value = obs[att_val] if att_val in obs else 0.0
-            return (value + 1.0) / (sum(obs.values()) + len(obs))
-        return 0.0
+        class_dist = self._att_dist_per_class[target_val]
+        value = class_dist[att_val]
+        try:
+            return value / sum(class_dist.values())
+        except ZeroDivisionError:
+            return 0.0
 
     def best_evaluated_split_suggestion(
         self, criterion, pre_split_dist, att_idx, binary_only
     ):
-        best_suggestion = None
-        att_values = sorted(
-            set(
-                [
-                    att_val
-                    for att_val_per_class in self._att_val_dist_per_class.values()
-                    for att_val in att_val_per_class
-                ]
-            )
-        )
+        best_suggestion = BranchFactory()
+
         if not binary_only:
             post_split_dist = self._class_dist_from_multiway_split()
             merit = criterion.merit_of_split(pre_split_dist, post_split_dist)
-            branch_mapping = {
-                attr_val: branch_id for branch_id, attr_val in enumerate(att_values)
-            }
-            best_suggestion = SplitSuggestion(
-                NominalMultiwayTest(att_idx, branch_mapping), post_split_dist, merit,
+
+            best_suggestion = BranchFactory(
+                merit,
+                att_idx,
+                sorted(self._att_values),
```

Comment:
Perhaps this answers my previous comment, but better ask.

Reply:
I think so! The access is made by key, but I use the ordered values of the categories to ensure the access is always performed consistently.